File: Hinweis Java/java/src/main/java/de/thm/mni/compilerbau/absyn/ParameterDefinition.java
package de.thm.mni.compilerbau.absyn;

import de.thm.mni.compilerbau.absyn.visitor.Visitor;
import de.thm.mni.compilerbau.table.Identifier;

/**
 * This class represents the local declaration of a parameter in SPL.
 * <p>
 * Parameters are defined in the parameter list of a procedure. They combine an {@link Identifier} with
 * a {@link TypeExpression}, expressing the parameters type and additionally have to store
 * whether the parameter is passed as a reference.
 * Parameters are only visible in the local scope of their procedure.
 */
public class ParameterDefinition extends Node {
    public final Identifier name;
    public final TypeExpression typeExpression;
    public final boolean isReference;

    /**
     * Creates a new node representing the declaration of a parameter in the head of a procedure.
     *
     * @param position       The position of the declaration in the source code.
     * @param name           The identifier of the defined parameter.
     * @param typeExpression The type expression used to express the parameters type.
     * @param isReference    A boolean value used to represent whether the parameter is passed as a reference.
     */
    public ParameterDefinition(Position position, Identifier name, TypeExpression typeExpression, boolean isReference) {
        super(position);
        this.name = name;
        this.typeExpression = typeExpression;
        this.isReference = isReference;
    }

    @Override
    public String toString() {
        return formatAst("ParameterDefinition", name, typeExpression, isReference);
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

File: Skeleton Kotlin/kotlin/src/main/kotlin/de/thm/mni/compilerbau/absyn/ParameterDefinition.kt
package de.thm.mni.compilerbau.absyn

import de.thm.mni.compilerbau.absyn.visitor.Visitor
import de.thm.mni.compilerbau.table.Identifier

/**
 * This class represents the local definition of a parameter in SPL.
 *
 * Parameters are declared in the parameter list of a procedure. They combine an [Identifier] with
 * a [TypeExpression], expressing the parameters type and additionally have to store
 * whether the parameter is passed as a reference.
 * Parameters are only visible in the local scope of their procedure.
 *
 * @param position       The position of the definition in the source code.
 * @param name           The identifier of the declared parameter.
 * @param typeExpression The type expression used to express the parameters type.
 * @param isReference    A boolean value used to represent whether the parameter is passed as a reference.
 */
class ParameterDefinition(
    position: Position,
    val name: Identifier,
    val typeExpression: TypeExpression,
    val isReference: Boolean
) : Node(position) {
    override fun accept(visitor: Visitor) = visitor.visit(this)

    override fun toString(): String = formatAst("ParameterDefinition", name, typeExpression, isReference)

}
